fixing namespace handling to work as originally intended

jmervine opened the pull request:
You should be able to set `FOO` or `NAMESPACE_FOO` and fetch via `Config.namespace_foo` when doing something like:

```ruby
namespace :namespace do
    optional :foo
end
```

Currently only setting `FOO` gives you a value.

cc @joshuatobin @reidmix

Commit:
fixing namespace handling to work as origionally intended

Commit:
fixing error handling and adding tests around it

Commit:
removing commented old handlers

jmervine commented on test/configerator_test.rb:
```diff
@@ -20,11 +20,11 @@ class TestConfigerator < Minitest::Test
+    test_namespace1: 'namespace1',
+    test_namespace2: 'namespace2',
 
-    ns1: 'ns1',
```
@reidmix using `test_ns{1,2}` and `ns{1,2}` is what masked the bug. =P

jmervine commented on lib/configerator/configerator.rb:
```diff
@@ -76,13 +77,63 @@ def cast(value, method)
+    instance_variable_set(:"@#{meth}", value)
+
+    if has_prefix?
+      instance_eval "def #{meth}; @#{meth} || (raise 'keys not set: \"#{pfxd}\" or \"#{orig}\"' unless #{error_on_load}) end"
```
@reidmix Because only one of the two keys are expected to be set, perhaps this should be `key not set:` and not `keys not set`, in spite of it referencing two keys. Thoughts?

lexelby replied:
I vote for singular.

lexelby commented on lib/configerator/configerator.rb:
```diff
@@ -76,13 +77,63 @@ def cast(value, method)
+    defined?(@prefix) && !@prefix.nil? && @prefix != ""
+  end
+
+  def prefixize_key(key)
```
IMO the "ize" is unnecessary: "prefix" is already a verb.  `prefix_key` should be sufficient.

lexelby commented on lib/configerator/configerator.rb:
```diff
@@ -76,13 +77,63 @@ def cast(value, method)
+
+              ENV[prefixed_key] || ENV[stringified_key] || default
+            else
+              # handle one possible keys
```
s/keys/key/

lexelby commented on test/configerator_test.rb:
```diff
@@ -63,10 +63,30 @@ def test_required_with_method
+      Config.required :test_missing
+    rescue => ex
+      assert_equal ex.message, 'key not found: "TEST_MISSING"'
+    end
```
Seems kind of repetitive, having to run `Config.required :test_missing` twice to test both the exception type and message.  Perhaps this?

```
begin
  Config.required :test_missing
rescue KeyError => ex
  assert_equal ex.message, 'key not found: "TEST_MISSING"'
else
  fail
end
```

jmervine replied:
Oh, that's nifty! Didn't know about `fail`.

Commit:
addressing comments from lexelby \o/

jmervine commented:
@lexelby All comments address. Thanks for the excellent feedback.

lexelby commented on test/configerator_test.rb:
```diff
@@ -64,14 +64,12 @@ def test_required_with_method
-
     begin
       Config.required :test_missing
     rescue => ex
```
Still need to specify `KeyError` here to make sure it raises the right exception.

lexelby commented on test/configerator_test.rb:
```diff
@@ -168,7 +166,7 @@ def test_namespace_required_missing_on_call
     rescue => ex
-      assert_equal ex.message, 'keys not set: "TEST_NAMESPACE99" or "NAMESPACE99"'
+      assert_equal ex.message, 'key not set: "TEST_NAMESPACE99" or "NAMESPACE99"'
     end
```
Is it possible to do the same begin/rescue/else thing here?

Commit:
addressing comments

lexelby commented:
:+1:

jmervine commented:
Based on a conversation with @reidmix, I'm going to update this so only `NAMESPACE_FOO` is supported and not `FOO`, when using namespaces by default. Disabling prefix when defining the  namespace will omit the requirement (and support) for the `NAMESPACE_`.

Commit:
adjust prefix to support to only support prefix with prefix true

Commit:
replacing minitest with rspec

Commit:
adding all ruby 2.x version to travis

Commit:
travis wants more specific versions

jmervine commented:
@reidmix Addressed comments from our previous discussion. Switch from minitest to rspec. I found the previous test suite hard to parse and it was leaving me feel less than confident in the tests. Also, because this is a gem, I updated travis to test against all 2.x versions of ruby.

Commit:
fixing versions in regression task

reidmix commented:
@jmervine awesome on rspec -- yeah i was super lacking on tests plus I'm not always so nimble on minitest.

reidmix commented on .travis.yml:
```diff
@@ -1,5 +1,9 @@
 sudo: false
 bundler_args: --without development
-script: make test
+script: make spec
```
is it weird we're using `make`?

jmervine replied:
Yeah, it tends to be my goto for simple things, I'll switch it to rake at some point, but perhaps not in this PR.

reidmix commented on lib/configerator/configerator.rb:
```diff
@@ -69,20 +70,39 @@ def array(method = nil)
   end
+
+  def stringify_key(key)
+    key = "#{@prefix}#{key}" if defined?(@prefix) && !@prefix.nil? && @prefix != ""
```
I believe you can safely do `if @prefix` to simplify and get what you're looking for. 

It'll return `nil` if its undefined.
If it's an empty string, it won't change the `key`.

jmervine replied:
I was trying to catch the `""` condition as well, but I think that's overkill. Good call here.

reidmix commented on lib/configerator/configerator.rb:
```diff
@@ -69,20 +70,39 @@ def array(method = nil)
     @processed << name
   end
+
+  def stringify_key(key)
```
I'd change the name here because its doing more than stringifying.  maybe `build_key`?

reidmix commented on lib/configerator/configerator.rb:
```diff
@@ -69,20 +70,39 @@ def array(method = nil)
+    key.to_s.upcase
+  end
+
+  def fetch_env(key, error_on_load: false, default: nil)
```
I don't think the `default` handling should be in this method, primarily because it's for one usecase (using `override`) and pushing it into this method complicates this method in subtle ways. 

The `error_on_load` functionality is no longer being executed quite the same way.  It's what `ENV.fetch` was doing when it raised an error if the key didn't exist, which is different from the value being `nil` (I'm realizing your change was done to help handle the case of applying the `default` values). 

My suggestion is to move it back to using `ENV.fetch` and let it do the error raising and not raise the error ourself here.   Then in `override` apply the default there (I'll make a note of it.)

jmervine replied:
Yeah, I was on the fence on this, I'll revisit.

reidmix commented on lib/configerator/configerator.rb:
```diff
@@ -1,36 +1,37 @@
+  end
+
+  def override(name, default, method=nil)
+    value = cast(fetch_env(name, default: default), method)
```
See my longer note on `fetch_env` -- I think we can simplify `fetch_env` if we apply the default inline here:

```ruby
    value = cast(fetch_env(name) || default, method)
```

reidmix commented on lib/configerator/configerator.rb:
```diff
@@ -1,36 +1,37 @@
-    # Hash#fetch raises a KeyError, Hash#[] doesn't
-    value = error_on_load ? ENV.fetch(name.to_s.upcase) : ENV[name.to_s.upcase]
-    value = cast(value, method)
+    value = fetch_env(name, error_on_load: error_on_load)
```
I'd probably make this look more like `optional` & `override` when it gets the value, for consistency:
```ruby
value = cast(fetch_env(name, error_on_load: error_on_load), method)
```

reidmix commented on lib/configerator/configerator.rb:
```diff
@@ -69,20 +70,39 @@ def array(method = nil)
+    instance_eval "def #{name}; @#{name} || (raise 'key not set: \"#{name.upcase}\"' unless #{error_on_load}) end", __FILE__, __LINE__
     instance_eval "def #{name}?; !!#{name} end", __FILE__, __LINE__
 
+    return unless @is_namespace
```
I don't think it hurts to have `@processed` get created if its not a namespace -- what you have here to check if it's for `@is_namespace` is more correct (for sure!) but I don't think you really need it.  the `@is_namespace` check doesn't really offer anything.  `@processed` gets cleared out every time `namespace` is called, so it stays tidy. 

Personally, I wouldn't worry about it and get rid of `@is_namespace`

reidmix commented on spec/configerator_spec.rb:
```diff
@@ -0,0 +1,275 @@
+require 'spec_helper'
+
+describe Configerator do
+  let(:config) { Module.new { extend Configerator } }
```
sweet!

reidmix commented on spec/configerator_spec.rb:
```diff
@@ -0,0 +1,275 @@
+      end
+    end
+
+    it "errors when missing required value" do
```
I'd also do a test for when `FOO` is not in the environment at all.

jmervine replied:
Setting it to `nil` is how you emulate it not being in the environment. If I don't remember it all all the `allow_any_instance_of` throws an error b/c something is called that's not expected. I might be able to do that by not stubbing the environment for that test, but then if someone happens to have `FOO` set, it will break the spec.

reidmix replied:
what if you don't mock anything?  that should do it right?

jmervine replied:
Yeah, oddly, initially I was getting errors with that, but I think that was in a previous iteration. I'm not sure what I changed to make it work, but that's what I'm doing now and it looks good.

reidmix commented on spec/configerator_spec.rb:
```diff
@@ -0,0 +1,275 @@
+      end
+    end
+
+    it "accepts missing value" do
```
I'd also do a test for when FOO is not in the environment at all.

reidmix commented on spec/configerator_spec.rb:
```diff
@@ -0,0 +1,275 @@
+    end
+
+    it "uses default with missing value" do
+      with_environment foo: nil do
```
I'd also do a test for when FOO is not in the environment at all.

reidmix commented on spec/configerator_spec.rb:
```diff
@@ -0,0 +1,275 @@
+
+          expect(config.foo).to  eq "method:bah"
+        end
+      end
```
👍

reidmix commented on spec/spec_helper.rb:
```diff
@@ -0,0 +1,17 @@
+  yield
+
+  unstub_environment
+end
```
I really like how this turned out in the tests, it looks really nice 👍

reidmix reviewed (changes requested):
This is looking really good -- I really like what you did with the tests. 

Most of my comments are around simplifying the code more than anything.

Commit:
removing @is_namespace per comments

Commit:
simplifying prefix check

Commit:
renaming stringify_key

Commit:
cleaning up fetch_env and specs around missing values

Commit:
reworking required a bit

Commit:
replacing make with rake

reidmix commented on lib/configerator/configerator.rb:
```diff
@@ -69,20 +69,34 @@ def array(method = nil)
   end
+
+  def build_key(key)
+    key = "#{@prefix}#{key}" if @prefix
```
i'm realizing we don't even need the `if @prefix` because if it's nil or blank, we'd never see it. 
Something to simplify for the future, not required to change now.

reidmix reviewed (approved):
LGTM 👍
